csharp/src/Microsoft.ML.OnnxRuntime/ProviderOptions.cs

```diff
@@ -0,0 +1,108 @@
+
+namespace Microsoft.ML.OnnxRuntime
+{
+    /// <summary>
+    /// Search Algorithm for Cudnn Conv.
+    /// </summary>
+    public enum OrtCudnnConvAlgoSearch
+    {
+        EXHAUSTIVE,  //!< expensive exhaustive benchmarking using cudnnFindConvolutionForwardAlgorithmEx
+        HEURISTIC,   //!< lightweight heuristic based search using cudnnGetConvolutionForwardAlgorithm_v7
+        DEFAULT,     //!< default algorithm using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM
+    }
+
+    /// <summary>
+    /// Provider options for CUDA EP.
+    /// </summary>
+    public struct OrtCUDAProviderOptions
+    {
+        public int device_id;                                   //!< cuda device with id=0 as default device.
+        public OrtCudnnConvAlgoSearch cudnn_conv_algo_search;   //!< cudnn conv algo search option
+        public UIntPtr gpu_mem_limit;                           //!< default cuda memory limitation to maximum finite value of size_t.
+        public int arena_extend_strategy;                       //!< default area extend strategy to KNextPowerOfTwo.
+        public int do_copy_in_default_stream;                   //!< Whether to do copies in the default stream or use separate streams.
+    }
+
+    /// <summary>
+    /// Holds provider options configuration for creating an InferenceSession.
+    /// </summary>
+    public class ProviderOptions : SafeHandle
+    {
+        internal IntPtr Handle
+        {
+            get
+            {
+                return handle;
+            }
+        }
+
+        #region Constructor and Factory methods
+
+        /// <summary>
+        /// Constructs an empty ProviderOptions
+        /// </summary>
+        public ProviderOptions()
+            : base(IntPtr.Zero, true)
+        {
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        /// <summary>
+        /// Get CUDA provider options with default setting.
+        /// </summary>
+        /// <returns> CUDA provider options instance.  </returns>
+        public static OrtCUDAProviderOptions GetDefaultCUDAProviderOptions()
+        {
+            OrtCUDAProviderOptions cuda_options;
+            cuda_options.device_id = 0;
+            cuda_options.cudnn_conv_algo_search = OrtCudnnConvAlgoSearch.EXHAUSTIVE;
+            if (IntPtr.Size == 8)
+            {
+                cuda_options.gpu_mem_limit = (UIntPtr)UInt64.MaxValue;
+            }
+            else
+            {
+                cuda_options.gpu_mem_limit = (UIntPtr)UInt32.MaxValue;
+            }
+            cuda_options.arena_extend_strategy = 0;
+            cuda_options.do_copy_in_default_stream = 1;
+
+            return cuda_options;
+        }
+        #endregion
+
+        #region Public Properties
+
+        /// <summary>
+        /// Overrides SafeHandle.IsInvalid
+        /// </summary>
+        /// <value>returns true if handle is equal to Zero</value>
+        public override bool IsInvalid { get { return handle == IntPtr.Zero; } }
```

Comment:
>IsInvalid [](start = 29, length = 9)

My understanding is that handle is never initialized and therefore is never valid. Is this ok?

Reply:
In constructor, IntPtr.Zero has been assigned to handle.
Therefore "handle == InPtr.Zero" will return true.

https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.-ctor?view=net-5.0
https://stackoverflow.com/questions/17645216/whats-the-purpose-of-invalidhandlevalue-in-safehandle

Reply:
Correct. I will always be invalid. What is the purpose of this class being SafeHandle?

---
In reply to: [612550263](https://github.com/microsoft/onnxruntime/pull/7315#discussion_r612550263) [](ancestors = 612550263)

Reply:
For future use. But agree with you that it's not needed to inherit from SafeHandle. Already made this class simple static class.